### PR TITLE
enable `#[deny(missing_docs)]`

### PR DESCRIPTION
**Pull request by nikomatsakis**
This supplies documentations for the plumbing module as well as various odds and ends in the `iter` module.

Also fixes https://github.com/rayon-rs/rayon/issues/414

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Hopefully I got all the links correct -- we don't have a link tester I guess.

File: src/iter/plumbing/mod.rs [hunk 1]
```diff
@@ -9,27 +12,84 @@ use super::IndexedParallelIterator;
 use std::cmp;
 use std::usize;
 
+/// The `ProducerCallback` trait is a kind of generic closure,
+/// [analogous to `FnOnce`][FnOnce]. See [the corresponding section in
+/// the plumbing README][r] for more details.
+///
+/// [r]: README.md#producer-callback
```
Comment (cuviper) [inline src/iter/plumbing/mod.rs]: Will this relative link still work on docs.rs and the like?

File: src/iter/plumbing/mod.rs [hunk 2]
```diff
@@ -242,6 +370,16 @@ pub fn bridge<I, C>(mut par_iter: I, consumer: C) -> C::Result
     }
 }
 
+/// This helper function is used to "connect" a producer and a
+/// consumer.  You may producer to call [`bridge`], which wraps this
```
Comment (cuviper) [inline src/iter/plumbing/mod.rs]: "You may producer to call `bridge`"?  Forgot a word?

**Comment by cuviper** [comment]
CI failed on missing docs for the unstable stuff.

File: src/iter/mod.rs [hunk 3]
```diff
@@ -1606,6 +1626,20 @@ pub trait IndexedParallelIterator: ParallelIterator {
 pub trait FromParallelIterator<T>
     where T: Send
 {
+    /// Creates an instance of the collection from the parallel iterator `par_iter`.
+    ///
+    /// If your collection is not naturally parallel, the easiest (and
+    /// fastest) way to do this is often to collect `par_iter` into a
+    /// [`LinkedList`] or other intermediate data structure and then
```
Comment (KodrAus) [inline src/iter/mod.rs]: Missing link?

File: src/iter/plumbing/mod.rs [hunk 4]
```diff
@@ -242,6 +370,16 @@ pub fn bridge<I, C>(mut par_iter: I, consumer: C) -> C::Result
...
+/// function. This function will draw items from `producer` and feed them to `consumer`,
+/// splitting and creating parallel tasks when needed.
+///
+/// This is useful when you are implementing your own parallel
+/// iterators: it is often used as the definition of the
+/// [`drive_unindexed`] or [`drive`] methods.
```
Comment (KodrAus) [inline src/iter/plumbing/mod.rs]: Missing links?

File: src/iter/plumbing/mod.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/iter/plumbing/mod.rs]: I *think* so...? That's why I made it relative. Could be wrong.

Comment (nikomatsakis) [reply src/iter/plumbing/mod.rs]: Oh, the README.md in particular .. hmm, maybe not. I could link back to the repo.

**Comment by nikomatsakis** [comment]
@cuviper @KodrAus addressed feedback (I believe)

**Comment by nikomatsakis** [comment]
bors r=cuviper

**Pull request merged**